lib/wherebot.rb

```diff
@@ -0,0 +1,142 @@
+require 'net/imap'
+require 'mail'
+
+class Wherebot
+  TRASH = "Trash".freeze
+  INBOX = "eff-where".freeze
+
+  class << self
+    def update_wheres(destructive: false)
+      imap = Net::IMAP.new(
+        ENV['INCOMING_MAIL_SERVER'], ENV['INCOMING_MAIL_PORT'], true
+      )
+
+      sign_into_wheremail(imap)
+
+      new_messages(imap).each do |id|
+        Wherebot::Message.new(imap, id, destructive).create
+      end
+
+      clean_up(imap, destructive)
+    end
+
+    private
+
+    def sign_into_wheremail(imap)
+      imap.login(ENV['WHEREBOT_USER_NAME'], ENV['WHEREBOT_PASSWORD'])
+      imap.examine(INBOX)
+    end
+
+    def new_messages(imap)
+      @new_messages ||= imap.search(["UNSEEN"])
+    end
+
+    def clean_up(imap, destructive)
+      imap.expunge if destructive
+      imap.logout
+      imap.disconnect
+    end
+  end
+
+  class Message
+    def initialize(imap, email_id, destroy = false)
+      @imap = imap
+      @id = email_id
+      @destroy = destroy
+    end
+
+    def create
+      wm = WhereMessage.find_or_create_by(
+        provenance: 'where@eff.org',
+        sent_at: DateTime.parse(headers.date),
+        user: User.find_by(
+          email: "#{headers.from[0].mailbox}@#{headers.from[0].host}"
+        )
+      )
+      if wm.update(body: body)
+        destroy_message
+      else
+        storage = ENV['WHERE_FAILURES'].try(:dup) || ''
+        storage << wm.to_json
+        ENV['WHERE_FAILURES'] = storage
+      end
+
+    end
+
+    def body
+      body = get_text
+      body = strip_quotes_and_signatures_from(body)
+      body = reject_encryption(body)
+      strip_junk_from(body)
+      body = remove_trailing_junk(body)
+      "#{nice_subject}\n#{body}".strip.chomp.force_encoding('UTF-8')
+    end
+
+    private
+
+    def headers
+      @headers ||= @imap.fetch(@id, "ENVELOPE")[0].attr["ENVELOPE"]
+    end
+
+    def full_body
+      @full_body ||= @imap.fetch(@id, 'BODY[TEXT]')[0].attr['BODY[TEXT]']
+    end
+
+    def get_text
+      body = full_body.split("Content-Type: text/")[1] || full_body
+      body = body.split("Content-Transfer-Encoding: ")[1] || body
```

Comment:
I see less good results from that.  Consider:
```
$ new_bodies = ids.last(5).map{|id| mail =  Mail.new(imap.fetch(id, 'BODY[TEXT]')[0].attr['BODY[TEXT]']); (mail.parts[0] || mail).body.decoded }
=> ["-- {sig} \n\n\n\n",
 "-- \n{sig}\n\n\n\n",
 "",
 "\n\n--\n{sig}\n\n\n",
 "\r\n-- \r\n{sig}\r\n\r\n"]
$ old_bodies = ids.last(5).map{|id| full_body = imap.fetch(id, 'BODY[TEXT]')[0].attr['BODY[TEXT]']; body = full_body.split("Content-Type: text/")[1] || full_body; body = body.split("Content-Transfer-Encoding: ")[1] || body; body }
=> ["Looks like for the third time this year, I've caught something. Heading\r\nhome to protect the heard\r\n\r\n-- \r\n{sig} \r\n\r\n\r\n\r\n",
 "auw\r\n\r\n-- \r\n{sig}\r\n\r\n\r\n\r\n",
 "Heading out around 4:30.\r\n\r\n\r\n\r\n",
 "7bit\r\n{sig}\r\n\r\n\r\n",
 "quoted-printable\r\n\r\n\r\n--=20\r\n{sig}\r\n\r\n\r\n"]
```
The first three items have useful bodies, but Mail loses them. Options!
* i'mma poke at it a little more to see if I can figure out why Mail loses the good bits and can be persuaded not to
* we could talk about it together, and combine our powers, like Voltron
* we could ship this as an MVP and keep working on it

Reply:
Do you see the same problem when you fetch the mails with this?

```ruby
imap.fetch(id, 'BODY.PEEK[]')[0].attr['BODY[]']
```

In other words `BODY.PEEK[]` instead of `BODY[TEXT]`.

Reply:
Ah good call.  😞 sorry i didn't catch that.  I've good too many loose ends on this project right now. I'll try to be more diligent! 😤